File: apps/reading/src/api/floatingDeepResearch.ts
/**
 * Floating deep research instances (pure client model).
 *
 * Operator vision: spin up a deep-research window from a reading/research
 * highlight; float it, open fullscreen, draft-merge into parent, fully merge,
 * or select multiple completed instances into a collective pack.
 *
 * Never invents live dispatch or executed merges. live_dispatched and
 * merge_executed are always false in this pure layer.
 */

export type FloatingResearchViewMode =
  | "floating"
  | "fullscreen"
  | "merged_draft"
  | "merged_full"
  | "collective";

export type FloatingResearchStatus =
  | "proposed"
  | "open"
  | "completed"
  | "closed";

export interface FloatingDeepResearchInstance {
  instance_id: string;
  parent_asset_id: string;
  highlight: string;
  prompt: string;
  view_mode: FloatingResearchViewMode;
  status: FloatingResearchStatus;
  /** Always false in pure layer — no provider dispatch. */
  live_dispatched: false;
  /** Always false in pure layer — merge is intent-only. */
  merge_executed: false;
  notes: string[];
  authority: "operator_spawn_only";
}

export interface MergeIntent {
  kind: "draft_merge" | "full_merge";
  instance_id: string;
  parent_asset_id: string;
  /** Always false — intent only, never auto-merge. */
  merge_executed: false;
  operator_ack: boolean;
  notes: string[];
}

export interface CollectivePackIntent {
  kind: "collective_pack";
  parent_asset_id: string;
  instance_ids: string[];
  /** Always false — pack proposal only. */
  pack_dispatched: false;
  notes: string[];
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function newInstanceId(parent: string, highlight: string): string {
  // Deterministic-enough client id without crypto dependency.
  const base = `${parent}:${highlight.slice(0, 48)}`;
  let h = 0;
  for (let i = 0; i < base.length; i++) {
    h = (h * 31 + base.charCodeAt(i)) | 0;
  }
  return `fdr_${Math.abs(h).toString(36)}_${parent.slice(0, 12)}`;
}

/**
 * Spawn a floating deep-research instance from a highlight.
 * Fail closed on gated highlights and empty fields.
 * live_dispatched is always false.
 */
export function spawnFloatingFromHighlight(input: {
  parent_asset_id: string;
  highlight: string;
  prompt?: string;
  gated: boolean;
  view_mode?: FloatingResearchViewMode;
}): FloatingDeepResearchInstance {
  if (typeof input.gated !== "boolean") {
    throw new Error(
      "gated must be an explicit boolean from highlight provenance (fail closed)",
    );
  }
  if (input.gated === true) {
    throw new Error("gated/withheld highlight cannot spawn floating deep research");
  }
  const parent = requireNonEmpty(input.parent_asset_id, "parent_asset_id");
  const highlight = requireNonEmpty(input.highlight, "highlight");
  const prompt =
    typeof input.prompt === "string" && input.prompt.trim()
      ? input.prompt.trim()
      : `Deep research on highlight from ${parent}`;
  const view_mode = input.view_mode ?? "floating";
  if (
    view_mode !== "floating" &&
    view_mode !== "fullscreen" &&
    view_mode !== "merged_draft" &&
    view_mode !== "merged_full" &&
    view_mode !== "collective"
  ) {
    throw new Error("view_mode invalid");
  }
  if (view_mode === "merged_full" || view_mode === "merged_draft") {
    throw new Error(
      "spawn cannot start already-merged; use proposeDraftMerge/proposeFullMerge",
    );
  }
  if (view_mode === "collective") {
    throw new Error("spawn cannot start as collective; use proposeCollectivePack");
  }

  const notes = [
    "spawned from highlight — pure client instance (no live dispatch)",
    "live_dispatched=false",
    "merge_executed=false",
  ];

  return {
    instance_id: newInstanceId(parent, highlight),
    parent_asset_id: parent,
    highlight,
    prompt,
    view_mode,
    status: "proposed",
    live_dispatched: false,
    merge_executed: false,
    notes,
    authority: "operator_spawn_only",
  };
}

export function setFloatingViewMode(
  instance: FloatingDeepResearchInstance,
  view_mode: FloatingResearchViewMode,
): FloatingDeepResearchInstance {
  if (!instance || typeof instance !== "object") {
    throw new Error("instance must be an object");
  }
  if (instance.live_dispatched !== false) {
    throw new Error("live_dispatched must be false (pure layer)");
  }
  if (instance.merge_executed !== false) {
    throw new Error("merge_executed must be false (pure layer)");
  }
  if (
    view_mode !== "floating" &&
    view_mode !== "fullscreen" &&
    view_mode !== "merged_draft" &&
    view_mode !== "merged_full" &&
    view_mode !== "collective"
  ) {
    throw new Error("view_mode invalid");
  }
  if (view_mode === "merged_draft" || view_mode === "merged_full") {
    throw new Error("use proposeDraftMerge/proposeFullMerge for merge modes");
  }
  if (view_mode === "collective") {
    throw new Error("use proposeCollectivePack for collective mode");
  }
  const notes = [...instance.notes, `view_mode → ${view_mode}`];
  return {
    ...instance,
    view_mode,
    status: instance.status === "proposed" ? "open" : instance.status,
    live_dispatched: false,
    merge_executed: false,
    notes,
  };
}

export function markFloatingCompleted(
  instance: FloatingDeepResearchInstance,
): FloatingDeepResearchInstance {
  if (instance.live_dispatched !== false) {
    throw new Error("live_dispatched must be false (pure layer)");
  }
  return {
    ...instance,
    status: "completed",
    live_dispatched: false,
    merge_executed: false,
    notes: [
      ...instance.notes,
      "marked completed by operator (no automatic provider completion)",
    ],
  };
}

/**
 * Propose draft merge into parent asset. Never executes merge.
 */
export function proposeDraftMerge(
  instance: FloatingDeepResearchInstance,
): MergeIntent {
  if (instance.live_dispatched !== false) {
    throw new Error("live_dispatched must be false");
  }
  if (
    instance.status !== "completed" &&
    instance.status !== "open" &&
    instance.status !== "proposed"
  ) {
    throw new Error("draft merge requires proposed, open, or completed instance");
  }
  return {
    kind: "draft_merge",
    instance_id: instance.instance_id,
    parent_asset_id: instance.parent_asset_id,
    merge_executed: false,
    operator_ack: false,
    notes: [
      "draft merge intent only — provisional combined document not written",
      "merge_executed=false",
    ],
  };
}

/**
 * Propose full merge. Requires explicit operator_ack; never executes.
 */
export function proposeFullMerge(
  instance: FloatingDeepResearchInstance,
  input: { operator_ack: boolean },
): MergeIntent {
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (input.operator_ack !== true) {
    throw new Error("full merge requires operator_ack=true (fail closed)");
  }
  if (instance.live_dispatched !== false) {
    throw new Error("live_dispatched must be false");
  }
  if (instance.status !== "completed") {
    throw new Error("full merge requires completed instance");
  }
  return {
    kind: "full_merge",
    instance_id: instance.instance_id,
    parent_asset_id: instance.parent_asset_id,
    merge_executed: false,
    operator_ack: true,
    notes: [
      "full merge intent only — parent asset not mutated in pure layer",
      "merge_executed=false",
    ],
  };
}

/**
 * Propose collective pack from multiple instances (same parent).
 * Never dispatches a collective research run.
 */
export function proposeCollectivePack(
  instances: FloatingDeepResearchInstance[],
): CollectivePackIntent {
  if (!Array.isArray(instances) || instances.length < 2) {
    throw new Error("collective pack requires at least 2 instances");
  }
  const parent = instances[0].parent_asset_id;
  const ids: string[] = [];
  for (const inst of instances) {
    if (!inst || typeof inst !== "object") {
      throw new Error("each instance must be an object");
    }
    if (inst.live_dispatched !== false) {
      throw new Error("live_dispatched must be false for all instances");
    }
    if (inst.parent_asset_id !== parent) {
      throw new Error("collective pack requires same parent_asset_id");
    }
    if (inst.status !== "completed" && inst.status !== "open") {
      throw new Error("collective pack requires open or completed instances");
    }
    ids.push(inst.instance_id);
  }
  // de-dupe preserving order
  const seen = new Set<string>();
  const unique = ids.filter((id) => {
    if (seen.has(id)) return false;
    seen.add(id);
    return true;
  });
  if (unique.length < 2) {
    throw new Error("collective pack requires at least 2 distinct instance_ids");
  }
  return {
    kind: "collective_pack",
    parent_asset_id: parent,
    instance_ids: unique,
    pack_dispatched: false,
    notes: [
      "collective pack intent only — no multi-agent dispatch",
      "pack_dispatched=false",
    ],
  };
}

export function formatFloatingSummary(
  instance: FloatingDeepResearchInstance,
): string {
  return (
    `id=${instance.instance_id} · mode=${instance.view_mode} · ` +
    `status=${instance.status} · live_dispatched=false · merge_executed=false`
  );
}

File: substrate/floating_deep_research.py
"""Floating deep research instances (pure, fail-closed).

Operator vision: spawn a deep-research window from a reading/research
highlight; float it, open fullscreen, draft-merge into parent, fully merge
(with ack), or select multiple instances into a collective pack.

This module never:
* dispatches live providers
* mutates parent assets
* invents merge completion

live_dispatched, merge_executed, and pack_dispatched are always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

ViewMode = Literal["floating", "fullscreen", "merged_draft", "merged_full", "collective"]
Status = Literal["proposed", "open", "completed", "closed"]

MAX_ID = 256
MAX_HIGHLIGHT = 8000
MAX_PROMPT = 4000


class FloatingDeepResearchError(ValueError):
    """Fail-closed validation for floating deep research."""


@dataclass(frozen=True)
class FloatingDeepResearchInstance:
    instance_id: str
    parent_asset_id: str
    highlight: str
    prompt: str
    view_mode: ViewMode
    status: Status
    live_dispatched: bool
    merge_executed: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "instance_id": self.instance_id,
            "parent_asset_id": self.parent_asset_id,
            "highlight": self.highlight,
            "prompt": self.prompt,
            "view_mode": self.view_mode,
            "status": self.status,
            "live_dispatched": False,
            "merge_executed": False,
            "notes": list(self.notes),
            "authority": "operator_spawn_only",
        }


@dataclass(frozen=True)
class MergeIntent:
    kind: Literal["draft_merge", "full_merge"]
    instance_id: str
    parent_asset_id: str
    merge_executed: bool
    operator_ack: bool
    notes: tuple[str, ...]

    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": self.kind,
            "instance_id": self.instance_id,
            "parent_asset_id": self.parent_asset_id,
            "merge_executed": False,
            "operator_ack": self.operator_ack,
            "notes": list(self.notes),
        }


@dataclass(frozen=True)
class CollectivePackIntent:
    kind: Literal["collective_pack"]
    parent_asset_id: str
    instance_ids: tuple[str, ...]
    pack_dispatched: bool
    notes: tuple[str, ...]

    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": "collective_pack",
            "parent_asset_id": self.parent_asset_id,
            "instance_ids": list(self.instance_ids),
            "pack_dispatched": False,
            "notes": list(self.notes),
        }


def _require_bool(value: object, *, field: str) -> bool:
    if not isinstance(value, bool):
        raise FloatingDeepResearchError(f"{field} must be an explicit boolean")
    return value


def _require_nonempty(value: object, *, field: str, max_len: int) -> str:
    if not isinstance(value, str):
        raise FloatingDeepResearchError(f"{field} must be a string")
    text = value.strip()
    if not text:
        raise FloatingDeepResearchError(f"{field} must be non-empty")
    if len(text) > max_len:
        raise FloatingDeepResearchError(f"{field} exceeds {max_len} chars")
    return text


def _instance_id(parent: str, highlight: str) -> str:
    base = f"{parent}:{highlight[:48]}"
    h = 0
    for ch in base:
        h = (h * 31 + ord(ch)) & 0xFFFFFFFF
    return f"fdr_{h:x}_{parent[:12]}"


def spawn_floating_from_highlight(
    *,
    parent_asset_id: object,
    highlight: object,
    gated: object,
    prompt: object | None = None,
    view_mode: object = "floating",
) -> FloatingDeepResearchInstance:
    """Spawn a pure client/server instance. Never live-dispatches."""
    if _require_bool(gated, field="gated") is True:
        raise FloatingDeepResearchError(
            "gated/withheld highlight cannot spawn floating deep research"
        )
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id", max_len=MAX_ID)
    hl = _require_nonempty(highlight, field="highlight", max_len=MAX_HIGHLIGHT)
    if prompt is None or (isinstance(prompt, str) and not prompt.strip()):
        prompt_s = f"Deep research on highlight from {parent}"
    else:
        prompt_s = _require_nonempty(prompt, field="prompt", max_len=MAX_PROMPT)
    if not isinstance(view_mode, str):
        raise FloatingDeepResearchError("view_mode must be a string")
    vm = view_mode.strip()
    if vm in ("merged_draft", "merged_full"):
        raise FloatingDeepResearchError(
            "spawn cannot start already-merged; use propose_draft_merge/propose_full_merge"
        )
    if vm == "collective":
        raise FloatingDeepResearchError(
            "spawn cannot start as collective; use propose_collective_pack"
        )
    if vm not in ("floating", "fullscreen"):
        raise FloatingDeepResearchError("view_mode invalid for spawn")

    notes = (
        "spawned from highlight — pure instance (no live dispatch)",
        "live_dispatched=false",
        "merge_executed=false",
    )
    return FloatingDeepResearchInstance(
        instance_id=_instance_id(parent, hl),
        parent_asset_id=parent,
        highlight=hl,
        prompt=prompt_s,
        view_mode=vm,  # type: ignore[arg-type]
        status="proposed",
        live_dispatched=False,
        merge_executed=False,
        notes=notes,
        authority="operator_spawn_only",
    )


def set_floating_view_mode(
    instance: FloatingDeepResearchInstance,
    view_mode: object,
) -> FloatingDeepResearchInstance:
    if not isinstance(instance, FloatingDeepResearchInstance):
        raise FloatingDeepResearchError("instance must be FloatingDeepResearchInstance")
    if instance.live_dispatched is not False or instance.merge_executed is not False:
        raise FloatingDeepResearchError("honesty flags must remain false")
    if not isinstance(view_mode, str):
        raise FloatingDeepResearchError("view_mode must be a string")
    vm = view_mode.strip()
    if vm in ("merged_draft", "merged_full"):
        raise FloatingDeepResearchError(
            "use propose_draft_merge/propose_full_merge for merge modes"
        )
    if vm == "collective":
        raise FloatingDeepResearchError("use propose_collective_pack for collective mode")
    if vm not in ("floating", "fullscreen"):
        raise FloatingDeepResearchError("view_mode invalid")
    status: Status = "open" if instance.status == "proposed" else instance.status
    return FloatingDeepResearchInstance(
        instance_id=instance.instance_id,
        parent_asset_id=instance.parent_asset_id,
        highlight=instance.highlight,
        prompt=instance.prompt,
        view_mode=vm,  # type: ignore[arg-type]
        status=status,
        live_dispatched=False,
        merge_executed=False,
        notes=instance.notes + (f"view_mode → {vm}",),
        authority="operator_spawn_only",
    )


def mark_floating_completed(
    instance: FloatingDeepResearchInstance,
) -> FloatingDeepResearchInstance:
    if not isinstance(instance, FloatingDeepResearchInstance):
        raise FloatingDeepResearchError("instance must be FloatingDeepResearchInstance")
    if instance.live_dispatched is not False:
        raise FloatingDeepResearchError("live_dispatched must be false")
    return FloatingDeepResearchInstance(
        instance_id=instance.instance_id,
        parent_asset_id=instance.parent_asset_id,
        highlight=instance.highlight,
        prompt=instance.prompt,
        view_mode=instance.view_mode,
        status="completed",
        live_dispatched=False,
        merge_executed=False,
        notes=instance.notes
        + ("marked completed by operator (no automatic provider completion)",),
        authority="operator_spawn_only",
    )


def propose_draft_merge(instance: FloatingDeepResearchInstance) -> MergeIntent:
    if not isinstance(instance, FloatingDeepResearchInstance):
        raise FloatingDeepResearchError("instance must be FloatingDeepResearchInstance")
    if instance.live_dispatched is not False:
        raise FloatingDeepResearchError("live_dispatched must be false")
    if instance.status not in ("proposed", "open", "completed"):
        raise FloatingDeepResearchError(
            "draft merge requires proposed, open, or completed instance"
        )
    return MergeIntent(
        kind="draft_merge",
        instance_id=instance.instance_id,
        parent_asset_id=instance.parent_asset_id,
        merge_executed=False,
        operator_ack=False,
        notes=(
            "draft merge intent only — provisional combined document not written",
            "merge_executed=false",
        ),
    )


def propose_full_merge(
    instance: FloatingDeepResearchInstance,
    *,
    operator_ack: object,
) -> MergeIntent:
    if not isinstance(instance, FloatingDeepResearchInstance):
        raise FloatingDeepResearchError("instance must be FloatingDeepResearchInstance")
    if _require_bool(operator_ack, field="operator_ack") is not True:
        raise FloatingDeepResearchError(
            "full merge requires operator_ack=true (fail closed)"
        )
    if instance.live_dispatched is not False:
        raise FloatingDeepResearchError("live_dispatched must be false")
    if instance.status != "completed":
        raise FloatingDeepResearchError("full merge requires completed instance")
    return MergeIntent(
        kind="full_merge",
        instance_id=instance.instance_id,
        parent_asset_id=instance.parent_asset_id,
        merge_executed=False,
        operator_ack=True,
        notes=(
            "full merge intent only — parent asset not mutated in pure layer",
            "merge_executed=false",
        ),
    )


def propose_collective_pack(
    instances: list[FloatingDeepResearchInstance] | tuple[FloatingDeepResearchInstance, ...],
) -> CollectivePackIntent:
    if not isinstance(instances, (list, tuple)) or len(instances) < 2:
        raise FloatingDeepResearchError("collective pack requires at least 2 instances")
    parent = instances[0].parent_asset_id
    ids: list[str] = []
    for inst in instances:
        if not isinstance(inst, FloatingDeepResearchInstance):
            raise FloatingDeepResearchError(
                "each instance must be FloatingDeepResearchInstance"
            )
        if inst.live_dispatched is not False:
            raise FloatingDeepResearchError(
                "live_dispatched must be false for all instances"
            )
        if inst.parent_asset_id != parent:
            raise FloatingDeepResearchError(
                "collective pack requires same parent_asset_id"
            )
        if inst.status not in ("open", "completed", "proposed"):
            raise FloatingDeepResearchError(
                "collective pack requires proposed, open, or completed instances"
            )
        ids.append(inst.instance_id)
    seen: set[str] = set()
    unique: list[str] = []
    for i in ids:
        if i in seen:
            continue
        seen.add(i)
        unique.append(i)
    if len(unique) < 2:
        raise FloatingDeepResearchError(
            "collective pack requires at least 2 distinct instance_ids"
        )
    return CollectivePackIntent(
        kind="collective_pack",
        parent_asset_id=parent,
        instance_ids=tuple(unique),
        pack_dispatched=False,
        notes=(
            "collective pack intent only — no multi-agent dispatch",
            "pack_dispatched=false",
        ),
    )


__all__ = [
    "CollectivePackIntent",
    "FloatingDeepResearchError",
    "FloatingDeepResearchInstance",
    "MergeIntent",
    "mark_floating_completed",
    "propose_collective_pack",
    "propose_draft_merge",
    "propose_full_merge",
    "set_floating_view_mode",
    "spawn_floating_from_highlight",
]
